de.metas.payment.esr/src/main/java/de/metas/payment/spi/impl/ESRCreaLogixStringParser.java

```diff
@@ -171,7 +171,11 @@ public IPaymentString parse(final Properties ctx, final String paymentTextOrigin
 		else
 		{
 			esrReferenceNoToMatch = referenceNumber.substring(7, 15);
```

Comment:
why do we have `esrReferenceNoToMatch = referenceNumber.substring(7, 15);` 
if `referenceNumberLength != 27`?

to me it seems that if `referenceNumberLength != 27` we do assume that `referenceNumberLength == 15`.
however, the case this very issue is about seems to a third case (length neither 27 nor 15).
if we have a "sortof" 27-digit string that somehow has just 23 digits, then `referenceNumber.substring(7, 15);` is probably not a great idea.

Please check out whats the issue, here and also write a testcase; extract the code under test into a dedicated method if required to have the test.